File: src/contests/cg/puzzle/fibonaccis-rabbit.ts
// 🎮 CodinGame Puzzle - fibonaccis-rabbit
// https://www.codingame.com/training/easy/fibonaccis-rabbit

const [f0, n]: number[] = readline().split(" ").map(Number)
const [a, b]: number[] = readline().split(" ").map(Number)

const f: bigint[] = new Array(n + 1).fill(0n)
f[0] = BigInt(f0)

for (let i = 1; i <= n; i++) {
  let sum = 0n
  for (let k = a; k <= b; k++) {
    if (i - k >= 0) sum += f[i - k]
  }
  f[i] = sum
}

console.log(f[n].toString())

File: src/contests/cg/puzzle/scala/fibonaccis-rabbit.scala
// 🎮 CodinGame Puzzle - fibonaccis-rabbit
// https://www.codingame.com/training/easy/fibonaccis-rabbit

object Solution {
  def main(args: Array[String]): Unit = {
    val Array(f0, n) = scala.io.StdIn.readLine().trim.split(" ").map(_.toInt)
    val Array(a, b) = scala.io.StdIn.readLine().trim.split(" ").map(_.toInt)
    // FN can exceed 2^63 (but is < 2^64), so accumulate with BigInt.
    val f = Array.fill[BigInt](n + 1)(BigInt(0))
    f(0) = BigInt(f0)
    for (i <- 1 to n) {
      var total = BigInt(0)
      for (k <- a to b) {
        if (i - k >= 0) total += f(i - k)
      }
      f(i) = total
    }
    println(f(n))
  }
}
